source/common/filesystem/kqueue/watcher_impl.cc

```diff
@@ -0,0 +1,175 @@
+#include "common/filesystem/watcher_impl.h"
+
+#include <sys/event.h>
+#include <sys/fcntl.h>
+#include <sys/types.h>
+
+#include "envoy/common/exception.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/event/file_event.h"
+
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+
+#include "event2/event.h"
+
+namespace Envoy {
+namespace Filesystem {
+
+WatcherImpl::WatcherImpl(Event::Dispatcher& dispatcher)
+    : queue_(kqueue()),
+      kqueue_event_(dispatcher.createFileEvent(queue_,
+                                               [this](uint32_t events) -> void {
+                                                 if (events & Event::FileReadyType::Read) {
+                                                   onKqueueEvent();
+                                                 }
+                                               },
+                                               Event::FileTriggerType::Edge,
+                                               Event::FileReadyType::Read)) {}
+
+WatcherImpl::~WatcherImpl() {
+  close(queue_);
+
+  for (const auto& entry : watches_) {
+    close(entry.first);
+  }
+  watches_.clear();
+}
+
+void WatcherImpl::addWatch(const std::string& path, uint32_t events, Watcher::OnChangedCb cb) {
+  FileWatchPtr watch = addWatch_(path, events, cb, false);
+  if (watch == nullptr) {
+    throw EnvoyException(fmt::format("invalid watch path {}", path));
+  }
+}
+
+WatcherImpl::FileWatchPtr WatcherImpl::addWatch_(const std::string& path, uint32_t events,
+                                                 Watcher::OnChangedCb cb, bool pathMustExist) {
+  bool watchingDir = false;
+  int watch_fd = open(path.c_str(), O_SYMLINK);
+  if (watch_fd == -1) {
+    if (pathMustExist) {
+      return nullptr;
+    }
+
+    size_t last_slash = path.rfind('/');
+    if (last_slash == std::string::npos) {
+      throw EnvoyException(fmt::format("invalid watch path {}", path));
+    }
+
+    std::string directory = path.substr(0, last_slash);
+    watch_fd = open(directory.c_str(), 0);
+    if (watch_fd == -1) {
+      return nullptr;
+    }
+
+    watchingDir = true;
+  }
+
+  FileWatchPtr watch(new FileWatch());
+  watch->fd_ = watch_fd;
+  watch->file_ = path;
+  watch->events_ = events;
+  watch->callback_ = cb;
+  watch->watchingDir_ = watchingDir;
+
+  int flags = NOTE_DELETE | NOTE_RENAME;
+  if (watchingDir) {
+    flags = NOTE_DELETE | NOTE_WRITE;
+  }
+
+  struct kevent event;
+  EV_SET(&event, watch_fd, EVFILT_VNODE, EV_ADD | EV_CLEAR, flags, 0,
+         reinterpret_cast<void*>(watch_fd));
+
+  if (kevent(queue_, &event, 1, NULL, 0, NULL) == -1) {
+    throw EnvoyException(
```

Comment:
watch->fd_ will leak here when you throw and in other cases. Can you add close() to the destructor of FileWatch, then you can make overall destruction simpler.